core/drivers/ls_sfp.c

```diff
@@ -0,0 +1,423 @@
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result ls_sfp_get_its(uint32_t *its)
+{
+	if (!sfp_regs) {
+		EMSG("SFP driver not initialized");
+		return TEE_ERROR_GENERIC;
+	}
+
+	if (!its) {
+		EMSG("Given buffer not initialized");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	*its = io_read32((vaddr_t)&sfp_regs->ospr0) & SFP_OSPR0_ITS;
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result ls_sfp_get_ouid(uint32_t index, uint32_t *ouid)
+{
+	if (!sfp_regs) {
+		EMSG("SFP driver not initialized");
+		return TEE_ERROR_GENERIC;
+	}
+
+	if (!ouid) {
+		EMSG("Given buffer not initialized");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	if (index >= ARRAY_SIZE(sfp_regs->ouidr)) {
+		EMSG("Index greater or equal to ouid: %"PRIu32" >= %"PRIu32,
+		     index, ARRAY_SIZE(sfp_regs->ouidr));
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	*ouid = io_read32((vaddr_t)&sfp_regs->ouidr[index]);
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result ls_sfp_get_sb(uint32_t *sb)
+{
+	if (!sfp_regs) {
+		EMSG("SFP driver not initialized");
+		return TEE_ERROR_GENERIC;
+	}
+
+	if (!sb) {
+		EMSG("Given buffer not initialized");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	*sb = io_read32((vaddr_t)&sfp_regs->sfpcr) & SFP_SFPCR_SB;
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result ls_sfp_get_srkh(uint32_t index, uint32_t *srkh)
+{
+	if (!sfp_regs) {
+		EMSG("SFP driver not initialized");
+		return TEE_ERROR_GENERIC;
+	}
+
+	if (!srkh) {
+		EMSG("Given buffer not initialized");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	if (index >= ARRAY_SIZE(sfp_regs->srkhr)) {
+		EMSG("Index greater or equal to srkhr: %"PRIu32" >= %"PRIu32,
+		     index, ARRAY_SIZE(sfp_regs->srkhr));
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	*srkh = io_read32((vaddr_t)&sfp_regs->srkhr[index]);
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result ls_sfp_set_debug_level(uint32_t dblev)
+{
+	uint32_t ospr1 = 0;
+
+	if (!sfp_regs) {
+		EMSG("SFP driver not initialized");
+		return TEE_ERROR_GENERIC;
+	}
+
+	if (!dblev) {
+		EMSG("Debug level can't be set to 0");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	ospr1 = io_read32((vaddr_t)&sfp_regs->ospr1);
+	if (ospr1 & SFP_OSPR1_DBLEV_MASK) {
+		EMSG("Debug level has already been fused");
+		return TEE_ERROR_SECURITY;
+	}
+
+	io_write32((vaddr_t)&sfp_regs->ospr1, ospr1 | dblev);
+
+	return ls_sfp_program_fuses();
+}
+
+TEE_Result ls_sfp_set_its_wp(void)
+{
+	uint32_t ospr0 = 0;
+
+	if (!sfp_regs) {
+		EMSG("SFP driver not initialized");
+		return TEE_ERROR_GENERIC;
+	}
+
+	ospr0 = io_read32((vaddr_t)&sfp_regs->ospr0);
+	if (ospr0 & (SFP_OSPR0_WP | SFP_OSPR0_ITS)) {
+		EMSG("SFP is already fused");
+		return TEE_ERROR_SECURITY;
+	}
+
+	ospr0 |= SFP_OSPR0_WP | SFP_OSPR0_ITS;
+	io_write32((vaddr_t)&sfp_regs->ospr0, ospr0);
+
+	return ls_sfp_program_fuses();
+}
+
+TEE_Result ls_sfp_set_ouid(uint32_t index, uint32_t ouid)
+{
+	if (!sfp_regs) {
+		EMSG("SFP driver not initialized");
+		return TEE_ERROR_GENERIC;
+	}
+
+	if (index >= ARRAY_SIZE(sfp_regs->ouidr)) {
+		EMSG("Index greater or equal to ouid: %"PRIu32" >= %"PRIu32,
```

Comment:
You have quite a few `EMSG()` which I think could be converted to `DMSG()` or simply discarded (letting the caller deal with the error).